test/e2e/syncer.go

```diff
@@ -0,0 +1,188 @@
+	"sigs.k8s.io/kubefed/pkg/controller/util"
+	"sigs.k8s.io/kubefed/pkg/kubefedctl"
+	kfenable "sigs.k8s.io/kubefed/pkg/kubefedctl/enable"
+	kfenableopts "sigs.k8s.io/kubefed/pkg/kubefedctl/options"
+	"sigs.k8s.io/kubefed/test/common"
+	"sigs.k8s.io/kubefed/test/e2e/framework"
+
+	. "github.com/onsi/ginkgo"
+)
+
+var _ = Describe("Sync controller", func() {
+	f := framework.NewKubeFedFramework("sync-controller-refresh")
+	tl := f.Logger()
+
+	var controller *federatedtypeconfig.Controller
+
+	BeforeEach(func() {
+		if framework.TestContext.RunControllers() {
+			var controllerFixture *framework.ControllerFixture
+			controllerFixture, controller = framework.NewFederatedTypeConfigControllerFixture(tl, f.ControllerConfig())
+			f.RegisterFixture(controllerFixture)
+		}
+	})
+
+	It("should be refreshed if the target CRD version changed for a federated CRD", func() {
+		if !framework.TestContext.RunControllers() {
+			framework.Skipf("sync controller refresh test is not supported for unmanaged control plane.")
+		}
+
+		By("Creating a CRD with version 1")
+		targetCrdKind := fmt.Sprintf("FedTarget-%s", uuid.New()[0:10])
+		targetAPIResource := metav1.APIResource{
+			Group:   kfenableopts.DefaultFederatedGroup,
+			Version: kfenableopts.DefaultFederatedVersion,
+
+			Kind:       targetCrdKind,
+			Name:       fedv1b1.PluralName(targetCrdKind),
+			Namespaced: true,
+		}
+		targetCrd := kfenable.CrdForAPIResource(targetAPIResource, nil, nil)
+		crdClient := apiextv1b1client.NewForConfigOrDie(f.KubeConfig())
+
+		targetCrd.Spec.Version = "v1"
+		createCrdForHost(f.Logger(), crdClient, targetCrd)
+		waitForTargetCrd(f.Logger(), f.KubeConfig(), targetAPIResource.Name, "v1")
+
+		By("Federating the CRD v1")
+		needCleanup := true
+		typeConfig := enableResource(f, &targetAPIResource, "v1", needCleanup)
+
+		By("Waiting for sync controller ready")
+		objectMeta := typeConfig.GetObjectMeta()
+		waitForSyncerVersion(tl, controller, objectMeta.Name, "v1")
+
+		By("Upgrading the CRD version from v1 to v2")
+		existingCrd, err := crdClient.CustomResourceDefinitions().Get(objectMeta.Name, metav1.GetOptions{})
+		if err != nil {
+			tl.Fatalf("Error retrieving target CRD %q: %v", objectMeta.Name, err)
+		}
+		existingCrd.Spec.Version = "v2"
+		existingCrd.Spec.Versions = []apiextv1b1.CustomResourceDefinitionVersion{
+			{
+				Name:    "v2",
+				Served:  true,
+				Storage: true,
+			},
+			{
+				Name:    "v1",
+				Served:  false,
+				Storage: false,
+			},
+		}
+		_, err = crdClient.CustomResourceDefinitions().Update(existingCrd)
+		if err != nil {
+			tl.Fatalf("Error updating target CRD version %q: %v", existingCrd.Spec.Version, err)
+		}
+
+		By("Federating the CRD v2")
+		needCleanup = false
+		typeConfig = enableResource(f, &targetAPIResource, "v2", needCleanup)
+
+		By("Waiting for refreshed sync controller ready")
+		objectMeta = typeConfig.GetObjectMeta()
+		waitForSyncerVersion(tl, controller, objectMeta.Name, "v2")
```

Comment:
How does this check ensure that the sync controller was restarted?

Reply:
When a sync controller is restart, its `version` changed. So, if `version`, e.g. "v2", matches controller version, it tells a sync controller with the specified version started up.

If `Generation` is used, I will try other method for detect.